bin/pyntlm_auth/handlers.py

```diff
@@ -0,0 +1,154 @@
+import ncache
+import re
+import hashlib
+import binascii
+import json
+import utils
+import ms_event
+import global_vars
+import rpc
+
+from samba import param, NTSTATUSError, ntstatus
+
+
+def format_response(nt_key_or_error_msg, error_code):
+    if error_code == 0:
+        return f'NT_KEY: {nt_key_or_error_msg}', HTTPStatus.OK
+
+    if error_code == ntstatus.NT_STATUS_WRONG_PASSWORD:
+        return nt_key_or_error_msg, HTTPStatus.UNAUTHORIZED
+
+    if error_code == ntstatus.NT_STATUS_NO_SUCH_USER:
+        return nt_key_or_error_msg, HTTPStatus.NOT_FOUND
+
+    if error_code == ntstatus.NT_STATUS_ACCOUNT_LOCKED_OUT or error_code == ntstatus.NT_STATUS_ACCOUNT_DISABLED:
+        return nt_key_or_error_msg, HTTPStatus.LOCKED
+
+    if error_code == global_vars.n_device_blocked:
+        return nt_key_or_error_msg, HTTPStatus.UNAUTHORIZED
+
+    return nt_key_or_error_msg, HTTPStatus.INTERNAL_SERVER_ERROR
+
+
+def ping_handler():
+    return "pong", HTTPStatus.OK
+
+
+def ntlm_connect_handler():
+    with global_vars.s_lock:
+        global_vars.s_reconnect_id = global_vars.s_connection_id
+
+    global_vars.s_secure_channel_connection, global_vars.s_machine_cred, global_vars.s_connection_id, error_code, error_message = rpc.get_secure_channel_connection()
+
+    if error_code == ntstatus.NT_STATUS_ACCESS_DENIED:
+        return "Test machine account failed. Access Denied", HTTPStatus.UNAUTHORIZED
+    if error_code != 0:
+        return "Error while establishing secure channel connection: " + error_message, HTTPStatus.INTERNAL_SERVER_ERROR
+
+    return "OK", HTTPStatus.OK
+
+
+def test_password_handler():
+    data = request.get_json()
+
+    if data is None:
+        return 'No JSON payload found in request', HTTPStatus.BAD_REQUEST
+    if 'password' not in data:
+        return 'Invalid JSON payload format, missing required key: password', HTTPStatus.UNPROCESSABLE_ENTITY
+
+    test_password = data['password'].strip()
+
+    if re.search(r'^[a-fA-F0-9]{32}$', test_password):
+        nt_hash = test_password
+    else:
+        nt4_digest = hashlib.new('md4', test_password.encode('utf-16le')).digest()
+        nt_hash = binascii.hexlify(nt4_digest).decode('utf-8')
+
+    global_vars.c_password = nt_hash
+
+    with global_vars.s_lock:
+        global_vars.s_reconnect_id = global_vars.s_connection_id
+
+    global_vars.s_secure_channel_connection, global_vars.s_machine_cred, global_vars.s_connection_id, error_code, error_message = rpc.get_secure_channel_connection()
+
+    if error_code == ntstatus.NT_STATUS_ACCESS_DENIED:
+        return "Test machine account failed. Access Denied", HTTPStatus.UNAUTHORIZED
+    if error_code != 0:
+        return "Error while establishing secure channel connection: " + error_message, HTTPStatus.INTERNAL_SERVER_ERROR
+
+    return "OK", HTTPStatus.OK
+
+
+def ntlm_auth_handler():
+    try:
+        required_keys = {'username', 'mac', 'request-nt-key', 'challenge', 'nt-response'}
```

Comment:
Yes. since mac address will be used for identifying a specific device, it's required. 
not sure if we can get a mac address in this case: "In the case of mschap for a VPN access there is no Calling-Station-Id attribute. We should have an option to have the mac and not having the mac."